### PR TITLE
PXC-3031: Added MTR testcase

### PR DESCRIPTION
**Pull request by dutow**

### CODE REVIEW

File: mysql-test/suite/galera/galera_2nodes.cnf [hunk 1]
```diff
@@ -25,6 +25,7 @@ wsrep_node_incoming_address=127.0.0.1:@mysqld.1.port
 innodb_autoinc_lock_mode=2
 
 innodb_flush_log_at_trx_commit=2
+wsrep_node_name=node1
```
Comment (kamil-holubicki) [inline mysql-test/suite/galera/galera_2nodes.cnf:28]: Will it not change the output of other tests?

File: scripts/sst_garbd_test.sh [hunk 2]
```diff
@@ -0,0 +1,33 @@
+#!/bin/bash
```
Comment (kamil-holubicki) [inline scripts/sst_garbd_test.sh]: This file is used for testing purposes only. Wouldn't it be better to keep it in mysql-test/std_data/?

File: mysql-test/suite/galera/t/galera_garbd_backup.test [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+--sleep 2 
+
+# Error: incorrect receiver script
+--error 2
+--exec `dirname $WSREP_PROVIDER`/../bin/garbd --sst="xtrabackup-v2:127.0.0.1:9999/xtrabackup_sst//1" --recv-script="not_sst_garbd_test 9999" --address "gcomm://127.0.0.1:$NODE_GALERAPORT_1" --group my_wsrep_cluster --options 'evs.inactive_timeout=PT6S;base_port=$NODE_GALERAPORT_3' > /tmp/garbd2.log 2>&1 
+SELECT VARIABLE_VALUE FROM performance_schema.global_status WHERE VARIABLE_NAME = 'wsrep_cluster_size';
+
+#--echo Restarting node #2 to satisfy MTR's end-of-test checks
+#--connection node_2
```
Comment (kamil-holubicki) [inline mysql-test/suite/galera/t/galera_garbd_backup.test]: commented out code

File: mysql-test/suite/galera/t/galera_garbd_backup.test [hunk 4]
```diff
@@ -0,0 +1,65 @@
+#
+# Test a single node with an arbitrator.  Startup a 2-node cluster, shut down node 2, and use it's ports
```
Comment (kamil-holubicki) [inline mysql-test/suite/galera/t/galera_garbd_backup.test]: Is this comment relevant? I can't see shutting down node 2 during the test.
Node's 3 port seems to be used instead.

File: mysql-test/suite/galera/t/galera_garbd_backup.test [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+--echo Starting garbd ...
+--exec `dirname $WSREP_PROVIDER`/../bin/garbd --sst="xtrabackup-v2:127.0.0.1:9999/xtrabackup_sst//1" --recv-script="sst_garbd_test 9999" --address "gcomm://127.0.0.1:$NODE_GALERAPORT_1" --group my_wsrep_cluster --options 'evs.inactive_timeout=PT6S;base_port=$NODE_GALERAPORT_3' > /tmp/garbd.log 2>&1 
+SELECT VARIABLE_VALUE FROM performance_schema.global_status WHERE VARIABLE_NAME = 'wsrep_cluster_size';
+
+--sleep 2 
+
+# Error: incorrect donor sst script
+--error 1
+--exec `dirname $WSREP_PROVIDER`/../bin/garbd --sst="xtrabackup-v3:127.0.0.1:9998/xtrabackup_sst//1" --recv-script="sst_garbd_test 9998" --address "gcomm://127.0.0.1:$NODE_GALERAPORT_1" --group my_wsrep_cluster --options 'evs.inactive_timeout=PT6S;base_port=$NODE_GALERAPORT_3' > /tmp/garbd2.log 2>&1
```
Comment (kamil-holubicki) [inline mysql-test/suite/galera/t/galera_garbd_backup.test]: wouldn't it be better to explicitly say something like --sst="nonexistent-script"?

File: mysql-test/suite/galera/t/galera_garbd_backup.test [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+# Error: incorrect donor sst script
+--error 1
+--exec `dirname $WSREP_PROVIDER`/../bin/garbd --sst="xtrabackup-v3:127.0.0.1:9998/xtrabackup_sst//1" --recv-script="sst_garbd_test 9998" --address "gcomm://127.0.0.1:$NODE_GALERAPORT_1" --group my_wsrep_cluster --options 'evs.inactive_timeout=PT6S;base_port=$NODE_GALERAPORT_3' > /tmp/garbd2.log 2>&1 
+SELECT VARIABLE_VALUE FROM performance_schema.global_status WHERE VARIABLE_NAME = 'wsrep_cluster_size';
+
+--sleep 2 
+
+# Error: incorrect receiver script
+--error 2
+--exec `dirname $WSREP_PROVIDER`/../bin/garbd --sst="xtrabackup-v2:127.0.0.1:9999/xtrabackup_sst//1" --recv-script="not_sst_garbd_test 9999" --address "gcomm://127.0.0.1:$NODE_GALERAPORT_1" --group my_wsrep_cluster --options 'evs.inactive_timeout=PT6S;base_port=$NODE_GALERAPORT_3' > /tmp/garbd2.log 2>&1
```
Comment (kamil-holubicki) [inline mysql-test/suite/galera/t/galera_garbd_backup.test]: As above, something like --recv-script="nonexistent_script"

File: mysql-test/suite/galera/t/galera_garbd_backup.test [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+# Depending on how much time it passes since the previous garbd start, this either fails or succeeds.
+# More than a few ms: the second garbd process knows abaut the SST transfer for the first, and waits until it is done
+# Minimal time: both garbd processes think that the donor sends data to them.
+# Second garbd process only notices the issue when the SST process finishes yet it didn't receive any data, and it exists with a failure.
+--exec `dirname $WSREP_PROVIDER`/../bin/garbd --sst="xtrabackup-v2:127.0.0.1:9998/xtrabackup_sst//1" --donor node1 --recv-script="sst_garbd_test 9998" --address "gcomm://127.0.0.1:$NODE_GALERAPORT_1" --group my_wsrep_cluster --options 'evs.inactive_timeout=PT6S;base_port=$NODE_GALERAPORT_3' > /tmp/garbd4.log 2>&1  || exit 0
+
+--connection node_1
+
+# Wait until the background garbd process completes to avoid shutdown assertion
```
Comment (kamil-holubicki) [inline mysql-test/suite/galera/t/galera_garbd_backup.test:55]: Can we use start_proc_in_background.inc to start garbd and then wait_proc_to_finish.inc?

File: mysql-test/suite/galera/t/galera_garbd_backup.test [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+#--echo Restarting node #2 to satisfy MTR's end-of-test checks
+#--connection node_2
+#--source include/start_mysqld.inc
+
+# Simulate two connection error 
+
+--exec `dirname $WSREP_PROVIDER`/../bin/garbd --sst="xtrabackup-v2:127.0.0.1:9999/xtrabackup_sst//1" --donor node1 --recv-script="sst_garbd_test 9999" --address "gcomm://127.0.0.1:$NODE_GALERAPORT_1" --group my_wsrep_cluster --options 'evs.inactive_timeout=PT6S;base_port=$NODE_GALERAPORT_3' > /tmp/garbd3.log 2>&1 & 
+SELECT VARIABLE_VALUE FROM performance_schema.global_status WHERE VARIABLE_NAME = 'wsrep_cluster_size';
+
+# Depending on how much time it passes since the previous garbd start, this either fails or succeeds.
```
Comment (kamil-holubicki) [inline mysql-test/suite/galera/t/galera_garbd_backup.test:47]: How do we know if we are testing problematic situation and not always following happy path when both garbd receive the backup? 
Is it possible to pause server on some debug_sync point to force problematic situation?

File: mysql-test/suite/galera/galera_2nodes.cnf (hunk 1 above)
Comment (dutow) [reply mysql-test/suite/galera/galera_2nodes.cnf:28]: No, it doesn't show up anywhere.

File: mysql-test/suite/galera/t/galera_garbd_backup.test (hunk 8 above)
Comment (dutow) [reply mysql-test/suite/galera/t/galera_garbd_backup.test:47]: It's not possible: the issue happens in garbd, which doesn't have debug sync functionality.

File: mysql-test/suite/galera/t/galera_garbd_backup.test (hunk 4 above)
Comment (dutow) [reply mysql-test/suite/galera/t/galera_garbd_backup.test]: Removed.

File: mysql-test/suite/galera/t/galera_garbd_backup.test (hunk 3 above)
Comment (dutow) [reply mysql-test/suite/galera/t/galera_garbd_backup.test]: Removed.

File: scripts/sst_garbd_test.sh (hunk 2 above)
Comment (dutow) [reply scripts/sst_garbd_test.sh]: Good idea, moved.

File: mysql-test/suite/galera/t/galera_garbd_backup.test (hunk 7 above)
Comment (dutow) [reply mysql-test/suite/galera/t/galera_garbd_backup.test:55]: Good idea, done.

**Review by kamil-holubicki — APPROVED** [review]

**Comment by dutow** [comment]
Closing this PR: testcase was added as part of another PR.

**Pull request closed without merging**